codegen/genMetadata.ts

```diff
@@ -0,0 +1,107 @@
+import * as M from "../frame_metadata/mod.ts";
+import { Decl, getPath, getRawCodecPath, makeDocComment, S } from "./utils.ts";
+
+export function genMetadata(metadata: M.Metadata, decls: Decl[]) {
+  const { tys, extrinsic, pallets } = metadata;
+
+  const isUnitVisitor = new M.TyVisitor<boolean>(tys, {
+    unitStruct: () => true,
+    wrapperStruct(_, inner) {
+      return this.visit(inner);
+    },
+    tupleStruct: () => false,
+    objectStruct: () => false,
+    option: () => false,
+    result: () => false,
+    never: () => false,
+    stringUnion: () => false,
+    taggedUnion: () => false,
+    array: () => false,
+    sizedArray: () => false,
+    primitive: () => false,
+    compact: () => false,
+    bitSequence: () => false,
+    circular: () => false,
+  });
+
+  decls.push({
+    path: "_metadata.extrinsic",
+    code: [
+      "export const extrinsic =",
+      S.object(
+        ["version", extrinsic.version],
+        ["extras", getExtrasCodec(extrinsic.signedExtensions.map((x) => [x.ident, x.ty]))],
+        [
+          "additional",
+          getExtrasCodec(extrinsic.signedExtensions.map((x) => [x.ident, x.additionalSigned])),
+        ],
+      ),
+    ],
+  });
+  for (const pallet of pallets) {
+    for (const entry of pallet.storage?.entries ?? []) {
+      decls.push({
+        path: `${pallet.name}.${entry.name}`,
+        code: [
+          makeDocComment(entry.docs),
+          `export const ${entry.name} =`,
+          S.object(
+            ["type", S.string(entry.type)],
+            ["modifier", S.string(entry.modifier)],
+            [
+              "hashers",
+              entry.type === "Map" ? JSON.stringify(entry.hashers) : "[]",
+            ],
+            [
+              "key",
+              entry.type === "Map"
+                ? entry.hashers.length === 1
+                  ? ["$.tuple(", getRawCodecPath(tys[entry.key]!), ")"]
+                  : getRawCodecPath(tys[entry.key]!)
+                : "[]",
+            ],
+            ["value", getRawCodecPath(tys[entry.value]!)],
+          ),
+        ],
+      });
+    }
+    if (pallet.calls) {
+      const ty = tys[pallet.calls.ty]! as M.Ty & M.UnionTyDef;
+      const isStringUnion = ty.members.every((x) => !x.fields.length);
```

Comment:
A pallet, all the callables of which accept no arguments? I suppose it's possible ;)